### PR TITLE
First pass at stateful executor + persistence layer

### PR DESCRIPTION
**Pull request by Rob-Johnson**
This builds out the interface for Persisters, responsible for serializing + deserializing events. This also includes a PoC FilePersister, which writes out and retrieves events from files on disk. This isn't suitable for production, but demonstrates usage of the interface. I'm working on a Persister that'll use a 'proper' datastore.

### CODE REVIEW

File: task_processing/interfaces/event.py [hunk 1]
```diff
@@ -1,10 +1,13 @@
+import datetime
+
 from pyrsistent import field
 from pyrsistent import PRecord
 
 # TODO: organize and explain these
 
 
 class Event(PRecord):
+    timestamp = field(type=datetime.datetime)
```
Comment (keymone) [inline task_processing/interfaces/event.py]: i'd prefer a simple int here (or float)

File: task_processing/plugins/persistence/file_persistence.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from pyrsistent import v
+
+from task_processing.interfaces.event import Event
+from task_processing.interfaces.event import json_deserializer
+from task_processing.interfaces.event import json_serializer
+from task_processing.interfaces.persistence.reader import Reader
+from task_processing.interfaces.persistence.writer import Writer
+
+
+class FilePersistence(Reader, Writer):
```
Comment (keymone) [inline task_processing/plugins/persistence/file_persistence.py]: i'm not convinced we need such elaborate class hierarchy

File: task_processing/plugins/persistence/file_persistence.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import json
+
+from pyrsistent import v
```
Comment (jolynch) [inline task_processing/plugins/persistence/file_persistence.py]: I don't think that persistence is a plugin?

File: task_processing/plugins/persistence/file_persistence.py (hunk 2 above)
Comment (jolynch) [inline task_processing/plugins/persistence/file_persistence.py]: I agree, I'd just have like a Persistence interface that has read (read_status) and write (write_event)

File: task_processing/plugins/stateful/stateful_executor.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        worker_thread = threading.Thread(target=worker, daemon=True)
+        worker_thread.start()
+
+    def run(self, task_config):
+        self.downstream_executor.run(task_config)
+
+    def kill(self, task_id):
+        return self.downstream_executor.kill(task_id)
+
+    def status(self, task_id):
```
Comment (jolynch) [inline task_processing/plugins/stateful/stateful_executor.py]: Is status going away with the new ```event_queue``` property?

File: task_processing/plugins/stateful/stateful_executor.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            result = event_queue.get()
+            persister.write(event=result)
+            on_process(result)
+            event_queue.task_done()
+
+    return subscribe_to_updates_for_task
+
+
+def put_to_outbound_queue(event, processed_queue):
+    processed_queue.put(event)
```
Comment (Rob-Johnson) [inline task_processing/plugins/stateful/stateful_executor.py]: fixed - there wasn't any need for this to be its own function

File: examples/cluster/playground/Dockerfile [hunk 6]
```diff
@@ -16,12 +16,12 @@ RUN cd /tmp && \
...
-RUN pip install --upgrade pip==9.0.1
-RUN pip install --upgrade virtualenv==15.1.0
+RUN pip3 install --upgrade pip==9.0.1
+RUN pip3 install --upgrade virtualenv==15.1.0
 
 # FIXME use actual virtualenv
-RUN pip install pymesos
-RUN pip install pyrsistent
+RUN pip3 install pymesos
+RUN pip3 install pyrsistent
```
Comment (Rob-Johnson) [inline examples/cluster/playground/Dockerfile]: done

**Pull request merged**